### DISCUSSION TITLE
tutorials/voyage-part3/04-schema-checks-in-a-supergraph

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#91] [course feedback]
# tutorials/voyage-part3/04-schema-checks-in-a-supergraph

https://www.apollographql.com/tutorials/voyage-part3/04-schema-checks-in-a-supergraph

**Comment by torressam333** [comment]
Hello, for this step in the Launch: "Operation checks validate that a schema's changes won't break any operations that existing clients are sending to the graph. These checks are available for both federated and non-federated graphs."

How exactly is this implemented under the hood? How can the Schema Checks know for certain all of the operations that an existing client is sending? Thanks!

**Reply by lizhennessy** [reply]
Hi there @torressam333! This is a great question. When it's connected to GraphOS, the router can provide us visibility about how the graph is actually being used through a great deal of telemetry and metrics. (Here's the [official docs article](https://www.apollographql.com/docs/graphos/metrics/sending-operation-metrics/) on the subject!) When the router boots up and starts receiving requests, it begins to report operation metrics to GraphOS.  When it resolves an operation, how many times a field is requested, latency and errors; all of these are reported back to GraphOS, creating a nice clean dashboard that we can keep an eye on.

Then, when we make a change (such as changing a field in our schema!), GraphOS can use all of the metrics it's collected to validate the change that we're about to make. This gives it the basis to tell us "whoa, wait stop! you're about to break something!" or give us the go-ahead when nothing is wrong. :) Hope that answered your question, but please let us know if there's anything else we can help with!

**Comment by stephan-buckmaster** [comment]
If I see it right the schema checks don't go into resolver.js changes. However, these can also break the system

**Reply by stephan-buckmaster** [reply]
Following the course further, I get the impression that build checks and operation checks also wouldn't catch bugs in resolvers.

**Reply by lizhennessy** [reply]
Hey there @stephan-buckmaster ! That's right. Build checks ensure that our schema changes can be composed into the supergraph without breaking (i.e. there are no syntax errors in the schema). Operation checks ensure that we're not removing or changing parts of the schema that have been involved in historical operations (hence, frontends might be relying upon them the way they are!). These mechanisms aren't concerned with how the resolver logic is built, because it can be built in a number of ways! However, if you're using TypeScript, there's a convenient GraphQL Code Generator tool that makes validating your resolver code against your schema very intuitive! If you'd like, check out our [TypeScript series](https://www.apollographql.com/tutorials/browse?languages=TypeScript&paths=backend) which goes deeper into this topic!

**Reply by stephan-buckmaster** [reply]
Ok thanks! And many thanks for all your other responses 😊

**Reply by lizhennessy** [reply]
Of course @stephan-buckmaster ! Happy to help with any questions you might have, keep 'em coming. :)
